test/morgan.js

```diff
@@ -1316,6 +1317,102 @@ describe('morgan.compile(format)', function () {
   })
 })
 
+describe('morgan.token(name, function)', function () {
+  it('should overwrite existing functions if called multiple times', function (done) {
+    function token (req, res, arg) {
+      return arg && url.parse(req.url)[arg]
+    }
+    var urlToken = morgan.url
+    morgan.token('url', token)
+    assert.equal(morgan.url, token)
+    morgan.token('url', urlToken)
+    done()
+  })
+
+  describe('should use the string `-` if return value of the token function is falsey', function () {
+    function test (v, done) {
+      morgan.token('ret', function ret (req, res, arg) {
+        return arg
+      })
+      var cb = after(2, function (err, res, line) {
+        if (err) return done(err)
+        assert.equal(line, '-')
+        done()
+      })
+
+      var stream = createLineStream(function (line) {
+        cb(null, null, line)
+      })
+
+      var server = createServer(':ret', { stream: stream }, function (req, res, next) {
+        next()
+      })
+
+      request(server)
+      .get('/')
+      .expect(200, cb)
+    }
+    [
+      0,
+      NaN,
+      false,
+      '',
+      null,
+      undefined
+    ].forEach(function (v) {
+      it('for ' + (String(v) || JSON.stringify(v)), function (done) {
+        test(v, done)
+      })
+    })
+  })
+
+  it('should not see empty brackets as an argument value', function (done) {
```

Comment:
From the earlier comment, this test is not testing tokens; it's testing the parsing of the format string. i.m.o. the ignoring of the `[]` can really be considered a bug, not a feature. Perhaps we should be it rather than engrain the bug into the design?